src/utils/dockerDirManager.ts

```diff
@@ -0,0 +1,303 @@
+
+/**
+ * Goes under the `auths` key in docker config.json.
+ */
+type AuthConfig = {
+  username?: string,
+  password?: string,
+  auth?: string,
+  email?: string,
+  serveraddress?: string,
+  identitytoken?: string,
+  registrytoken?: string,
+}
+
+/**
+ * The parts of a docker config.json file that concern Rancher Desktop.
+ */
+type PartialDockerConfig = {
+  auths?: Record<string, AuthConfig>,
+  credsStore?: string,
+  credHelpers?: Record<string, string>,
+  currentContext?: string,
+}
+
+/**
+ * Manages everything under the docker CLI config directory (except, at
+ * the time of writing, docker CLI plugins).
+ */
+export default class DockerDirManager {
+  protected readonly dockerDirPath: string;
+  protected readonly dockerContextDirPath: string;
+  /**
+   * Path to the 'rancher-desktop' docker context file.  The parent directory
+   * is the SHA256 hash of the docker context name ('rancher-desktop'), per the
+   * docker convention.
+   */
+  protected readonly dockerContextPath: string;
+  protected readonly dockerConfigPath: string;
+  protected readonly defaultDockerSockPath = '/var/run/docker.sock';
+  protected readonly contextName = 'rancher-desktop';
+
+  /**
+   * @param dockerDirPath The path to the directory containing docker CLI config.
+   */
+  constructor(dockerDirPath: string) {
+    this.dockerDirPath = dockerDirPath;
+    this.dockerContextDirPath = path.join(this.dockerDirPath, 'contexts', 'meta');
+    this.dockerContextPath = path.join(this.dockerContextDirPath,
+      'b547d66a5de60e5f0843aba28283a8875c2ad72e99ba076060ef9ec7c09917c8', 'meta.json');
+    this.dockerConfigPath = path.join(this.dockerDirPath, 'config.json');
+    console.debug(`Created new DockerDirManager to manage dir: ${ this.dockerDirPath }`);
+  }
+
+  /**
+   * Gets the docker CLI config.json file as an object.
+   */
+  protected async readDockerConfig(): Promise<PartialDockerConfig> {
+    try {
+      const rawConfig = await fs.promises.readFile(this.dockerConfigPath, { encoding: 'utf-8' });
+
+      return JSON.parse(rawConfig);
+    } catch (error: any) {
+      if (error.code !== 'ENOENT') {
+        throw error;
+      }
+
+      return {};
+    }
+  }
+
+  /**
+   * Writes the docker CLI config.json file.
+   * @param config An object that is the config we want to write.
+   */
+  protected async writeDockerConfig(config: PartialDockerConfig): Promise<void> {
+    const rawConfig = JSON.stringify(config);
+
+    await fs.promises.mkdir(this.dockerDirPath, { recursive: true });
+    await fs.promises.writeFile(this.dockerConfigPath, rawConfig, { encoding: 'utf-8' });
+  }
+
+  /**
+   * Read the docker configuration, and return the docker socket in use by the
+   * current context.  If the context is invalid, return the default socket
+   * location.
+   * @param currentContext Docker's current context, as set in the configs.
+   */
+  protected async getCurrentDockerSocket(currentContext?: string): Promise<string> {
```

Comment:
Is this darwin/linux only? `unix://` wouldn't make sense on Windows.

Reply:
Assertion added.